src/pytorch_lightning/utilities/migration/utils.py

```diff
@@ -139,6 +146,7 @@ def _set_legacy_version(checkpoint: _CHECKPOINT, version: str) -> None:
     checkpoint.setdefault("legacy_pytorch-lightning_version", version)
 
 
-def _should_upgrade(checkpoint: _CHECKPOINT, target: str) -> bool:
+def _should_upgrade(checkpoint: _CHECKPOINT, target: str, max_version: Optional[str] = None) -> bool:
     """Returns whether a checkpoint qualifies for an upgrade when the version is lower than the given target."""
-    return Version(_get_version(checkpoint)) < Version(target)
+    is_lte_max_version = max_version is None or Version(target) <= Version(max_version)
+    return Version(_get_version(checkpoint)) < Version(target) and is_lte_max_version
```

Comment:
Reuse variable and shortcut if already False

```suggestion
    target_version = Version(target)
    is_lte_max_version = max_version is None or target_version <= Version(max_version)
    return is_lte_max_version and Version(_get_version(checkpoint)) < target_version
```